Updates to the architecture Registry section, and other misc

acnodal-tc opened the pull request:
This is triggered by the conversation at
https://cloud-native.slack.com/archives/CHSKJ4849/p1663793722639479

The main improvement from this patch is better clarity on the
different types of registry, and how other components interact with
the registry.

caboteria committed:
Architecture introduction tweaks

* Make the architecture venn diagram less enormous. IMHO it doesn't
really say much, but this way it doesn't take up most of the page
above the fold.
* There used to be two "A Network Service is..." sentences. I removed
one.
* Other text tweaks.

Signed-off-by: Toby Cabot <toby@acnodal.io>

caboteria committed:
Update the architecture registry section

This is triggered by the conversation at
https://cloud-native.slack.com/archives/CHSKJ4849/p1663793722639479

The main improvement from this patch is better clarity on the
different types of registry, and how other components interact with
the registry.

Signed-off-by: Toby Cabot <toby@acnodal.io>

caboteria committed:
Misc architecture page tweaks.

While I was editing the architecture page to update the Registry
section I noticed some other fixes.

Signed-off-by: Toby Cabot <toby@acnodal.io>

acnodal-tc commented:
...and I've now set up magit to add "signed-off-by" by default so hopefully there will be less force-pushing in my life.

edwarnicke commented on content/docs/concepts/architecture.md:
```diff
@@ -92,19 +90,23 @@ to 'refresh' the {{<color "#6A00FF" >}}vWire{{< /color >}}.  If a {{<color "#6A0
-## Registries
+## Registry
+
+The Network Service Registry (NSR) component tracks Network Services and Network Service {{<color "#008A00" >}}Endpoints{{</color>}}. It can use either in-memory volatile storage or Kubernetes custom resource storage.
```
So first: thank you.  This is a vast improvement.

You have correctly described the existing registry implementations.  I'm not sure if this is the right place or not, but it may be worth noting that one can fairly trivially write additional implementations with the backing store of your choice, because the 'Registry API' is *just* a GRPC API.

acnodal-tc replied:
I pushed a commit that adds links to the gRPC definition and a few other tweaks.

edwarnicke commented on content/docs/concepts/architecture.md:
```diff
@@ -92,19 +90,23 @@ to 'refresh' the {{<color "#6A00FF" >}}vWire{{< /color >}}.  If a {{<color "#6A0
+The Network Service Registry (NSR) component tracks Network Services and Network Service {{<color "#008A00" >}}Endpoints{{</color>}}. It can use either in-memory volatile storage or Kubernetes custom resource storage.
 
-As with any other Mesh, Network Service Mesh has Network Service Registries (NSR) in which Network Services and Network Service {{<color "#008A00" >}}Endpoint{{< /color >}}s are registered
+Clients and Endpoints don't connect directly to the Registry; they connect to the Network Service Manager which is a proxy for the registry. This simplifies NSM's internal data flows since Clients and Endpoints need to connect to the Manager anyway.
```
Your observation is absolutely true about the K8s implementation.  The indirection though is not a requirement in NSM.

Typically speaking, NSM utilizes a 'chain of responsibility' pattern.   (see [slides starting here](https://docs.google.com/presentation/d/154iByYvSI-g10RZRg6_xO0xBo-zpJopf/edit#slide=id.p16) ).

The chain of responsibility involves a 'proxy' for the NSM API (both the Request/Close/Monitor API and also for the Registry APIs) that assist in getting your vWire to/from an environment.  The NSMgr/Forwarder in K8s plays that role for a 'Node'.  

The local Pod literally can't handle getting a vWire to itself: it's not privleged, it has no access to getting a tunnel in or out of the Node.  The NSMgr/Forwarder act as a proxy to provide those functions to the Pod tranparently.  So the Pod reasons about what it wants, local mechnisms like kernel interface or memif interface, and leaves reasoning about how to get in or out of the Node to the NSMgr/Forwarder.

The same plays out getting in or out of the K8s Cluster.  The NSMgrproxy handles issues like translating between NodeIP and ExternalNodeIP etc.

This architecture makes it possible to further handle getting to/from other 'containing environments' like an on-prem DC, cloud VPC, etc by introducing a new proxy into the chain.

Please also note that a proxy may handle control plane, dataplane, or both.  The NSMgr/Forwarder end up touching both the controlplane and dataplane - changing control plane messaging to replace the 'local mechanism' of kernel or memif with a 'remote mechanism' like vxlan or wireguard (among other changes) - but also creating part of the dataplane as well (the cross connect in the Forwarder between the kernel interface or memif dropped into the Pod and the tunnel leaving the Node.

NSMgrproxy by contrast is only really proxying the control plane - it jiggers the IP address parameters of whatever remote mechanism is in use, but not setting up anything that will touch the data plane.

This ability to separate or combine control plane and data plane at each hop is... extremely handy :)

edwarnicke replied:
Oh... and one other thing to be aware about wrt Registries: each Registry serves a domain.   While NSM generally has a 'Registry' domain local to each K8s cluster as currently implemented, you can also 'floating registries' not associated with any cluster.  To be consumable outside a local environment, Registries use SRV records for a DNS domain.  So if I want a Registry for the sales.example.com domain I would simply provide an SRV record pointing to its registry and then any workload that can then request a network service ('myns' for example) as myns@sales.example.com and, presuming the registry is willing to talk to the workload and the NSEs/Proxies are willing to accept the Request... get a vWire strung to them.

The same is true for NSEs.  An NSE can register with any Registry Domain that will have it (ie, authz).

acnodal-tc replied:
This is great info, thank you! Much of this is covered already but we need to add sections for the NSMgr and Forwarder to explain their roles. I think that might be out of scope of this PR, though.

edwarnicke replied:
Definitely out of scope of this PR.  Also, the NSMgr/Forwarder are conceptually local to the K8s implementation.  Other environments may (or may not) have those particular components.

acnodal-tc replied:
On that note, I just pushed a small change to explain about the general/k8s-specific architecture.

edwarnicke commented:
@acnodal-tc Overall this is a vast improvement.  I added a few comments adding color.  I look forward to seeing what you will do with them :)

caboteria committed:
Address review comments

* Link to the Registry gRPC proto to indicate that the current
implementations aren't the only possible ones.

Also a few other improvements:

* Link to cmd-registry source repos
* Rearrange Registry sections to better match the "Key Concepts" flow

Signed-off-by: Toby Cabot <toby@acnodal.io>

caboteria committed:
Link to Network Service API proto

... and use "gRPC" on this page since that's how Google spells it.

Signed-off-by: Toby Cabot <toby@acnodal.io>

caboteria committed:
Rearrange architecture "Key Concepts" sections

Moved the Endpoints section higher so we define both Clients and
Endpoints before vWires since vWires depend on the other concepts.

Signed-off-by: Toby Cabot <toby@acnodal.io>

acnodal-tc commented:
I pushed a few updates that hopefully help.

caboteria committed:
Arch page: explain generic vs. k8s-specific and link to k8s page

Signed-off-by: Toby Cabot <toby@acnodal.io>

denis-tingaikin reviewed (approved):
@edwarnicke Do you have any comments on this?

edwarnicke commented on content/docs/concepts/architecture.md:
```diff
@@ -48,63 +64,47 @@ The contract of a {{<color "#6A00FF" >}}vWire{{< /color >}} is:
 
-A {{<color "#008A00" >}}Network Service Endpoint{{< /color >}} ({{<color "#008A00" >}}NSE{{< /color >}} or {{<color "#008A00" >}}Endpoint{{< /color >}}) provides one or more Network Services.  It registers with the registry a list
-of Network Services (by name) that it provides, and the 'destination labels' it is advertising for each Network Service.
+Clients and Endpoints don't connect directly to the Registry; they connect to the [Network Service Manager](https://github.com/networkservicemesh/cmd-nsmgr) which acts as a proxy for the registry. This simplifies NSM's internal data flows since Clients and Endpoints need to connect to the Manager anyway.
```
```suggestion
In K8s Clients and Endpoints don't connect directly to the Registry; they connect to the [Network Service Manager](https://github.com/networkservicemesh/cmd-nsmgr) which acts as a proxy for the registry. This simplifies NSM's internal data flows since Clients and Endpoints need to connect to the Manager anyway.
```

edwarnicke replied:
It might make more sense to move this into the K8s tabs, as its a feature of the K8s implementation of NSM, not a fundamental part of the base architecture.
